File: python/canvastekk_workflow_sdk/workflow/builder.py
"""
Workflow Builder

Fluent API for constructing workflow definitions with built-in START/END nodes.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from canvastekk_workflow_sdk.workflow.models import (
    EdgeType,
    WorkflowDefinitionNode,
    WorkflowDefinitionSpec,
    WorkflowEdgeDefinition,
)

if TYPE_CHECKING:
    pass


class WorkflowBuilder:
    """Fluent builder for workflow definitions.

    Example::

        spec = (
            WorkflowBuilder()
            .add_start("start", outputs=["point_cloud"])
            .add_node("segment", slug="segmentation-v1.0.0", inputs={"method": "dbscan"})
            .add_end("end")
            .connect("start", "segment", from_output="point_cloud", to_input="input_file")
            .connect("segment", "end", from_output="instances", to_input="result")
            .build()
        )
    """

    def __init__(self) -> None:
        self._nodes: list[WorkflowDefinitionNode] = []
        self._edges: list[WorkflowEdgeDefinition] = []
        self._node_ids: set[str] = set()
        self._has_start = False

    def _check_duplicate(self, node_id: str) -> None:
        if node_id in self._node_ids:
            raise ValueError(f"Duplicate node ID: '{node_id}'")

    def add_start(
        self,
        node_id: str = "start",
        *,
        outputs: list[str] | dict[str, Any] | None = None,
        config_schema: dict[str, Any] | None = None,
        workflow_node_id: str | None = None,
    ) -> WorkflowBuilder:
        if self._has_start:
            raise ValueError("Workflow already has a START node. Only one is allowed.")
        self._check_duplicate(node_id)

        cs: dict[str, Any] | None = config_schema
        if cs is None and outputs is not None:
            if isinstance(outputs, list):
                props = {name: {"type": "string"} for name in outputs}
            else:
                props = outputs
            cs = {
                "type": "object",
                "properties": props,
            }

        self._nodes.append(
            WorkflowDefinitionNode(
                id=node_id,
                slug="__start__",
                name="START",
                inputs={},
                config_schema=cs,
                workflow_node_id=workflow_node_id,
            )
        )
        self._node_ids.add(node_id)
        self._has_start = True
        return self

    def add_end(
        self,
        node_id: str = "end",
        *,
        workflow_node_id: str | None = None,
    ) -> WorkflowBuilder:
        self._check_duplicate(node_id)
        self._nodes.append(
            WorkflowDefinitionNode(
                id=node_id,
                slug="__end__",
                name="END",
                inputs={},
                workflow_node_id=workflow_node_id,
            )
        )
        self._node_ids.add(node_id)
        return self

    def add_node(
        self,
        node_id: str,
        *,
        slug: str | None = None,
        name: str | None = None,
        inputs: dict[str, Any] | None = None,
        version: str | None = None,
        workflow_node_id: str | None = None,
        config_schema: dict[str, Any] | None = None,
    ) -> WorkflowBuilder:
        if slug in ("__start__", "__end__"):
            raise ValueError(
                f"Cannot use reserved slug '{slug}'. Use add_start() or add_end() instead."
            )
        self._check_duplicate(node_id)
        self._nodes.append(
            WorkflowDefinitionNode(
                id=node_id,
                slug=slug,
                name=name,
                inputs=inputs or {},
                version=version,
                workflow_node_id=workflow_node_id,
                config_schema=config_schema,
            )
        )
        self._node_ids.add(node_id)
        return self

    def connect(
        self,
        from_node: str,
        to_node: str,
        *,
        from_output: str = "",
        to_input: str = "",
        edge_type: EdgeType = EdgeType.DEFAULT,
        condition: str | None = None,
    ) -> WorkflowBuilder:
        if from_node not in self._node_ids:
            raise ValueError(f"Unknown source node: '{from_node}'")
        if to_node not in self._node_ids:
            raise ValueError(f"Unknown target node: '{to_node}'")

        self._edges.append(
            WorkflowEdgeDefinition(
                from_node=from_node,
                to_node=to_node,
                from_output=from_output,
                to_input=to_input,
                edge_type=edge_type,
                condition=condition,
            )
        )
        return self

    def build(self, *, validate: bool = True) -> WorkflowDefinitionSpec:
        from canvastekk_workflow_sdk.workflow.validation import validate as validate_graph

        spec = WorkflowDefinitionSpec(
            nodes=list(self._nodes),
            edges=list(self._edges),
        )

        if validate:
            result = validate_graph(spec)
            if not result.is_valid:
                from canvastekk_workflow_sdk.exceptions import WorkflowValidationError
                raise WorkflowValidationError(
                    f"Workflow validation failed: {'; '.join(result.errors)}"
                )

        return spec

File: typescript/src/workflow/builder.ts
import type { EdgeType, WorkflowEdgeDefinition, WorkflowDefinitionNode, WorkflowDefinitionSpec } from "./models.js";
import { validate } from "./validation.js";

/**
 * Builder for creating workflow specifications.
 */
export class WorkflowBuilder {
  private _nodes: WorkflowDefinitionNode[] = [];
  private _edges: WorkflowEdgeDefinition[] = [];
  private _nodeIds: Set<string> = new Set();
  private _hasStart = false;

  /**
   * Creates a new workflow builder.
   */
  constructor() {}

  /**
   * Checks if a node ID is already used.
   * @param nodeId - Node ID to check
   * @throws {Error} If node ID is duplicate
   */
  private checkDuplicate(nodeId: string): void {
    if (this._nodeIds.has(nodeId)) {
      throw new Error(`Duplicate node ID: '${nodeId}'`);
    }
  }

  /**
   * Adds a START node to the workflow.
   * @param nodeId - Node ID (default: "start")
   * @param opts - Start node options
   * @returns This builder instance for chaining
   * @throws {Error} If START node already exists
   */
  addStart(
    nodeId = "start",
    opts?: {
      outputs?: string[] | Record<string, unknown>;
      workflowNodeId?: string;
      configSchema?: Record<string, unknown>;
    },
  ): this {
    if (this._hasStart) {
      throw new Error("Workflow already has a START node. Only one is allowed.");
    }
    this.checkDuplicate(nodeId);

    let configSchema = opts?.configSchema;
    if (configSchema === undefined && opts?.outputs) {
      const outputs = opts.outputs;
      const props = Array.isArray(outputs)
        ? Object.fromEntries(outputs.map((n) => [n, { type: "string" }]))
        : outputs;
      configSchema = { type: "object", properties: props };
    }

    this._nodes.push({
      id: nodeId,
      slug: "__start__",
      name: "START",
      inputs: {},
      workflow_node_id: opts?.workflowNodeId ?? null,
      config_schema: configSchema ?? null,
    });
    this._nodeIds.add(nodeId);
    this._hasStart = true;
    return this;
  }

  /**
   * Adds an END node to the workflow.
   * @param nodeId - Node ID (default: "end")
   * @param opts - End node options
   * @returns This builder instance for chaining
   */
  addEnd(nodeId = "end", opts?: { workflowNodeId?: string }): this {
    this.checkDuplicate(nodeId);
    this._nodes.push({
      id: nodeId,
      slug: "__end__",
      name: "END",
      inputs: {},
      workflow_node_id: opts?.workflowNodeId ?? null,
    });
    this._nodeIds.add(nodeId);
    return this;
  }

  /**
   * Adds a workflow node.
   * @param nodeId - Unique node ID
   * @param opts - Node configuration options
   * @returns This builder instance for chaining
   * @throws {Error} If node ID is duplicate or slug is reserved
   */
  addNode(
    nodeId: string,
    opts: {
      slug?: string;
      name?: string;
      inputs?: Record<string, unknown>;
      version?: string;
      workflowNodeId?: string;
      configSchema?: Record<string, unknown>;
    },
  ): this {
    if (opts.slug === "__start__" || opts.slug === "__end__") {
      throw new Error(
        `Cannot use reserved slug '${opts.slug}'. Use addStart() or addEnd() instead.`,
      );
    }
    this.checkDuplicate(nodeId);
    this._nodes.push({
      id: nodeId,
      slug: opts.slug ?? null,
      name: opts.name ?? null,
      inputs: opts.inputs ?? {},
      version: opts.version ?? null,
      workflow_node_id: opts?.workflowNodeId ?? null,
      config_schema: opts?.configSchema ?? null,
    });
    this._nodeIds.add(nodeId);
    return this;
  }

  /**
   * Connects two nodes with an edge.
   *
   * @param fromNode - Source node ID
   * @param toNode - Target node ID
   * @param opts - Connection options
   * @returns This builder instance for chaining
   * @throws {Error} If source or target node not found, or if connecting a node to itself
   */
  connect(
    fromNode: string,
    toNode: string,
    opts?: {
      fromOutput?: string;
      toInput?: string;
      edgeType?: EdgeType;
      condition?: string;
    },
  ): this {
    if (!this._nodeIds.has(fromNode)) {
      throw new Error(`Unknown source node: '${fromNode}'`);
    }
    if (!this._nodeIds.has(toNode)) {
      throw new Error(`Unknown target node: '${toNode}'`);
    }
    if (fromNode === toNode) {
      throw new Error(`Self-loop detected: cannot connect '${fromNode}' to itself`);
    }
    this._edges.push({
      id: crypto.randomUUID(),
      from_node: fromNode,
      to_node: toNode,
      from_output: opts?.fromOutput ?? "",
      to_input: opts?.toInput ?? "",
      edge_type: opts?.edgeType ?? "default",
      condition: opts?.condition ?? null,
    });
    return this;
  }

  /**
   * Builds the workflow specification.
   * @param opts - Build options
   * @returns Workflow specification
   * @throws {Error} If validation fails
   */
  build(opts?: { validate?: boolean }): WorkflowDefinitionSpec {
    const spec: WorkflowDefinitionSpec = {
      nodes: [...this._nodes],
      edges: [...this._edges],
      metadata: {},
    };

    if (opts?.validate !== false) {
      const result = validate(spec);
      if (!result.isValid) {
        throw new Error(`Workflow validation failed: ${result.errors.join("; ")}`);
      }
    }

    return spec;
  }
}
